### PR TITLE
Reusable table

### PR DESCRIPTION
**Pull request by Mario-k4**
#281

### CODE REVIEW

**Comment by edichoska** [comment]
@Mario-k4 Please resolve the conflicts! Also please look into this, the table rows are not showing on the table 
image

File: app/admin-dashboard/page.tsx [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+/* eslint-disable-next-line no-unused-vars */
+const AdminDashboard = () => {
+  const { isModalOpen, showModal, closeModal } = useModal();
+
+  return (
+    <>
+      <SearchAndFilter />
+      <div>
+        <button type="button" onClick={showModal}>
+          Open Modal
```
Comment (edichoska) [inline app/admin-dashboard/page.tsx:38]: I believe these were added for testing purposes. We don't need them on this page. Remove them.

**Comment by edichoska** [comment]
cc: @aleksandarmicev please look into the issue as well

**Comment by edichoska** [comment]
@Mario-k4 @aleksandarmicev Can we make this the highest priority? It is still not working properly

File: components/reusable-components/reusable-table/ReusableTable.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        <tbody>
+          {data.map((item) => (
+            <TableRowComponent
+              key={item.id}
+              data={item}
+              isChecked={checkedId === item.id}
+              onCheckboxChange={handleCheckboxChange}
+            />
+          ))}
+        </tbody>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/ReusableTable.tsx:75]: We should display a message if there is no data at the moment.

File: components/reusable-components/reusable-table/TableHead.tsx [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+      <tr>
+        {headers.map((header) => (
+          <th
+            key={header as string}
+            className={style.tableHeaderCell}
+            onClick={() => onSort(header)}
+          >
+            {displayNames[header]}
+            {sortField === header && (
+              <span className={style.sortArrow}>{sortOrder === 'asc' ? ' ▲' : ' ▼'}</span>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableHead.tsx:31]: @aleksandarmicev the arrows should be visible all the time. By default the data should be sorted DESC, from newest to oldest.

**Comment by edichoska** [comment]
Consider either:
1. Adding min-height to table body, and displaying message: No data available
2. If no data, don't display the whole table. Right now we are showing only the thead

File: components/reusable-components/reusable-table/FetchData.tsx [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+  last_name: string;
+  role: string;
+}
+
+const url = 'http://localhost:3001/users';
+
+const headers = {
+  'Content-Type': 'application/json',
+  Accept: 'application/json',
+};
```
Comment (edichoska) [inline components/reusable-components/reusable-table/FetchData.tsx:13]: we should fetch the blog posts here, not users. Please do not hardcore urls.
For security reasons we should keep the URLs in the .env file.
In the code we will use them as :
const url = process.env.NEXT_PUBLIC_API_BASE_URL + '/content/blog-posts'!;
or use dummy data so we can see and test the table

Comment (edichoska) [reply components/reusable-components/reusable-table/FetchData.tsx:13]: Also use more descriptive names for all functions in this file

File: components/reusable-components/reusable-table/ReusableTable.tsx [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        <tbody>
+          {sortedData.map((item) => (
+            <TableRowComponent
+              key={item.id}
+              data={item}
+              isChecked={checkedId === item.id}
+              onCheckboxChange={handleCheckboxChange}
+            />
+          ))}
+        </tbody>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/ReusableTable.tsx:82]: what if we don't receive data? We should always take all possible UI states into consideration and make sure components behave consistently and correctly throughout.

File: app/admin-dashboard/page.tsx [hunk 6]
```diff
@@ -8,24 +10,35 @@ import {
...
+  role: 'Role',
+};
 
+/* eslint-disable-next-line no-unused-vars */
 const AdminDashboard = () => {
   const { isModalOpen, showModal, closeModal } = useModal();
 
   return (
     <>
       <SearchAndFilter />
```
Comment (edichoska) [inline app/admin-dashboard/page.tsx:35]: why is the search and filter component displaying a table?

File: components/reusable-components/reusable-table/ActionDropdown.tsx [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  return (
+    <div className={style.dropdown} ref={dropdownRef}>
+      <button type="button" onClick={() => setIsOpen(!isOpen)} disabled={isDisabled}>
+        Actions
+      </button>
+      {isOpen && (
+        <ul className={style.dropdownMenu}>
+          <li className={style.dropdownItem}>View</li>
+          <li className={style.dropdownItem}>Edit</li>
+          <li className={style.dropdownItem}>Delete</li>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/ActionDropdown.tsx:34]: to make the action dropdown truly reusable, lets not hardcode these values

File: components/reusable-components/reusable-table/TableHead.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            key={header as string}
+            className={style.tableHeaderCell}
+            onClick={() => onSort(header)}
+          >
+            {displayNames[header]}
+            <span className={style.sortArrow}> {getSortOrder(header) === 'asc' ? ' ▲' : ' ▼'}</span>
+          </th>
+        ))}
+      </tr>
+    </thead>
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableHead.tsx:43]: lets add two empty fields
the first field should be empty, also the last one

**Comment by edichoska** [comment]
@aleksandarmicev The sort is not working! Can you please take a look?

**Comment by aleksandarmicev** [comment]
> @aleksandarmicev The sort is not working! Can you please take a look?

With the latest updates, sorting in ascending and descending order is now working. If anything else needs to be changed, we are here to do it.

**Comment by edichoska** [comment]
When I hover over the Table head it changes the cursor, indicating I can type. Please look into that as well. And resolve the conflicts.
@Mario-k4 @aleksandarmicev

File: components/module-components/SearchAndFilter/DisplayNames.tsx [hunk 9]
```diff
@@ -34,7 +34,7 @@ const DisplayNames = ({ filterValue, selectedRoles }: DisplayNamesProps) => {
         });
         setNames(filteredNames);
       } catch (error) {
-        console.error('Error fetching data:', error);
+        // console.error('Error fetching data:', error);
```
Comment (edichoska) [inline components/module-components/SearchAndFilter/DisplayNames.tsx:37]: remove commented out console.log

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+import style from './tableRowComponent.module.scss';
+import ActionDropdown from './ActionDropdown';
+import { UserData } from './ReusableTable';
+
+interface DropdownItem {
+  id: string;
+  label: string;
+}
+
+interface TableRowComponentProps<T extends UserData> {
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:13]: we shouldn't extend UserData, this way the Table Row is not reusable

File: components/reusable-components/reusable-table/TableRowComponent.tsx [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+
+  return (
+    <tr className={style.rowComponent}>
+      <td aria-label="Checkbox">
+        <input type="checkbox" checked={isChecked} onChange={handleCheckboxChange} />
+      </td>
+      <td>{data.first_name}</td>
+      <td>{data.last_name}</td>
+      <td>{data.role}</td>
+      <td className={style.actionCell} aria-label="Actions">
```
Comment (edichoska) [inline components/reusable-components/reusable-table/TableRowComponent.tsx:42]: we should map over the data

File: components/reusable-components/reusable-table/TableRowComponent.tsx (hunk 10 above)
Comment (edichoska) [reply components/reusable-components/reusable-table/TableRowComponent.tsx:13]: We should also consider the following scenario:
what if we don't want to show every entry from the data in the table.
Example we are displaying the blog posts. Each blog post contains: id, title, description, slug, imageUrl, author, createdAt, updatedAt
But in the table we want to just display: title, description, imageUrl, author.

Comment (aleksandarmicev) [reply components/reusable-components/reusable-table/TableRowComponent.tsx:13]: Need some time to implement this scenario, I hope will be very soon...

Comment (edichoska) [reply components/reusable-components/reusable-table/TableRowComponent.tsx]: you would need another key in the TableRowProps interface that would be an []
in the ReusableTable you will pass the headers in that prop

**Review by edichoska — APPROVED** [review]
great work guys 🚀  @aleksandarmicev @Mario-k4

**Pull request merged**